# 对数几率回归

## 实验内容
1. 使用对数几率回归完成垃圾邮件分类问题和Dota2结果预测问题。
2. 计算十折交叉验证下的精度(accuracy)，查准率(precision)，查全率(recall)，F1值。

## 评测指标  
1. 精度
2. 查准率
3. 查全率
4. F1

# 1. 读取数据

In [7]:
import numpy as np

In [8]:
spambase = np.loadtxt('data/spambase/spambase.data', delimiter = ",")
dota2results = np.loadtxt('data/dota2Dataset/dota2Train.csv', delimiter=',')

# 2. 导入模型

In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_predict

# 3. 提取数据

这里的spamx和dota2x包含了数据集内所有的特征

In [10]:
spamx = spambase[:, :57]
spamy = spambase[:, 57]

dota2x = dota2results[:, 1:]
dota2y = dota2results[:, 0]

# 4. 训练并预测

请你完成两个模型使用全部特征的训练与预测，并将预测结果存储起来

In [11]:
# YOUR CODE HERE

spam_model = LogisticRegression()
# spam_model.fit(spamx, spamy)
# spam_prediction = spam_model.predict(spamx)
spam_prediction = cross_val_predict(spam_model, spamx, spamy)

dota2_model = LogisticRegression()
# dota2_model.fit(dota2x, dota2y)
# dota2_prediction = dota2_model.predict(dota2x)
dota2_prediction = cross_val_predict(dota2_model, dota2x, dota2y)


# print(spam_prediction)
# print(dota2_prediction)

# 可视化
# import matplotlib.pyplot as plt
# %matplotlib inline

# plt.subplot(121)
# plt.plot(spamx, spamy, '.', label = 'training data')
# plt.plot(spamx, spam_model.predict(spamx), '-', label = 'prediction')
# plt.xlabel('spamX')
# plt.ylabel('spamY')
# plt.title("spam set")
# plt.legend()

# plt.subplot(122)
# plt.plot(dota2x, dota2y, '.', label = 'training data')
# plt.plot(dota2x, dota2_model.predict(dota2x), '-', label = 'prediction')
# plt.xlabel('dota2X')
# plt.ylabel('dota2Y')
# plt.title("dota2 set")
# plt.legend()


# 5. 评价指标的计算

请你计算两个模型的四项指标

In [16]:
# YOUR CODE HERE
spam_accuracy_score = accuracy_score(spam_prediction, spamy)
spam_precision_score = precision_score(spam_prediction, spamy)
spam_recall_score = recall_score(spam_prediction, spamy)
spam_f1_score = f1_score(spam_prediction, spamy)

print(spam_accuracy_score, spam_precision_score, spam_recall_score, spam_f1_score)


dota2_accuracy_score = accuracy_score(dota2_prediction, dota2y)
dota2_precision_score = precision_score(dota2_prediction, dota2y)
dota2_recall_score = recall_score(dota2_prediction, dota2y)
dota2_f1_score = f1_score(dota2_prediction, dota2y)

print(dota2_accuracy_score, dota2_precision_score, dota2_recall_score, dota2_f1_score)



0.903933927407 0.88306674021 0.874385581649 0.878704720088
0.598445763627 0.676171538682 0.606431092808 0.639405276523


###### 双击此处填写

数据集|精度|查准率|查全率|F1
-|-|-|-|-
spambase | 0.903933927407 | 0.88306674021 | 0.874385581649 | 0.878704720088
dota2Results | 0.598445763627 | 0.676171538682 | 0.606431092808 | 0.639405276523

# 6. 选做：尝试对特征进行变换、筛选、组合后，训练模型并计算十折交叉验证后的四项指标

In [13]:
# YOUR CODE HERE




###### 双击此处填写
1. 模型1的处理流程：
2. 模型2的处理流程：
3. 模型3的处理流程:

模型|数据集|精度|查准率|查全率|F1
-|-|-|-|-|-
模型1 | 数据集 | 0.0 | 0.0 | 0.0 | 0.0
模型2 | 数据集 | 0.0 | 0.0 | 0.0 | 0.0
模型3 | 数据集 | 0.0 | 0.0 | 0.0 | 0.0